File: src/test/java/com/codereviews/hotel/HotelQuoteServiceTest.java
package com.codereviews.hotel;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertSame;

import java.math.BigDecimal;
import java.time.LocalDate;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class HotelQuoteServiceTest {
    private HotelQuoteService service;

    @BeforeEach
    void setUp() {
        service = new HotelQuoteService();
    }

    @Test
    void quotesOneNightWeekdayStay() {
        HotelQuote quote = service.quote(new HotelQuoteRequest(
                LocalDate.of(2026, 7, 1),
                LocalDate.of(2026, 7, 2),
                1,
                2,
                new BigDecimal("100.00"),
                "USD",
                null
        ));

        assertEquals(1, quote.nights());
        assertEquals(1, quote.rooms());
        assertEquals(2, quote.guests());
        assertEquals("USD", quote.currency());
        assertEquals(new BigDecimal("100.00"), quote.subtotal());
        assertEquals(new BigDecimal("0.00"), quote.discount());
        assertEquals(new BigDecimal("12.00"), quote.tax());
        assertEquals(new BigDecimal("112.00"), quote.total());
    }

    @Test
    void appliesWeekendPricingAndLoyaltyDiscount() {
        HotelQuote quote = service.quote(new HotelQuoteRequest(
                LocalDate.of(2026, 7, 3),
                LocalDate.of(2026, 7, 5),
                1,
                2,
                new BigDecimal("100.00"),
                "USD",
                "gold"
        ));

        assertEquals(2, quote.nights());
        assertEquals(new BigDecimal("236.00"), quote.subtotal());
        assertEquals(new BigDecimal("16.52"), quote.discount());
        assertEquals(new BigDecimal("26.34"), quote.tax());
        assertEquals(new BigDecimal("245.82"), quote.total());
    }

    @Test
    void addsExtraGuestFeesWhenRoomCapacityIsExceeded() {
        HotelQuote quote = service.quote(new HotelQuoteRequest(
                LocalDate.of(2026, 7, 6),
                LocalDate.of(2026, 7, 7),
                1,
                3,
                new BigDecimal("120.00"),
                "USD",
                "none"
        ));

        assertEquals(new BigDecimal("145.00"), quote.subtotal());
        assertEquals(new BigDecimal("162.40"), quote.total());
    }

    @Test
    void usesCachedQuoteForRepeatedRequests() {
        HotelQuoteRequest request = new HotelQuoteRequest(
                LocalDate.of(2026, 7, 6),
                LocalDate.of(2026, 7, 7),
                1,
                1,
                new BigDecimal("90.00"),
                "USD",
                "none"
        );

        HotelQuote firstQuote = service.quote(request);
        HotelQuote secondQuote = service.quote(request);

        assertSame(firstQuote, secondQuote);
    }
}

File: test/hotelQuoteService.test.js
import test from "node:test";
import assert from "node:assert/strict";

import { clearQuoteCache, quoteHotelStay } from "../src/hotelQuoteService.js";

test("quotes a one-night weekday stay", () => {
  clearQuoteCache();

  const quote = quoteHotelStay({
    checkIn: "2026-07-01",
    checkOut: "2026-07-02",
    rooms: 1,
    guests: 2,
    nightlyRate: 100,
    currency: "USD"
  });

  assert.deepEqual(quote, {
    nights: 1,
    rooms: 1,
    guests: 2,
    currency: "USD",
    subtotal: 100,
    discount: 0,
    tax: 12,
    total: 112
  });
});

test("applies weekend pricing and loyalty discount", () => {
  clearQuoteCache();

  const quote = quoteHotelStay({
    checkIn: "2026-07-03",
    checkOut: "2026-07-05",
    rooms: 1,
    guests: 2,
    nightlyRate: 100,
    loyaltyTier: "gold"
  });

  assert.equal(quote.nights, 2);
  assert.equal(quote.subtotal, 236);
  assert.equal(quote.discount, 16.52);
  assert.equal(quote.tax, 26.34);
  assert.equal(quote.total, 245.82);
});

test("adds extra guest fees when room capacity is exceeded", () => {
  clearQuoteCache();

  const quote = quoteHotelStay({
    checkIn: "2026-07-06",
    checkOut: "2026-07-07",
    rooms: 1,
    guests: 3,
    nightlyRate: 120
  });

  assert.equal(quote.subtotal, 145);
  assert.equal(quote.total, 162.4);
});

test("uses the cached quote for repeated requests", () => {
  clearQuoteCache();

  const request = {
    checkIn: "2026-07-06",
    checkOut: "2026-07-07",
    rooms: 1,
    guests: 1,
    nightlyRate: 90
  };

  const firstQuote = quoteHotelStay(request);
  const secondQuote = quoteHotelStay(request);

  assert.equal(firstQuote, secondQuote);
});
